core/controller/src/main/scala/whisk/core/controller/SequenceActions.scala

```diff
@@ -0,0 +1,399 @@
+                    val atomicActionCnt = wskActivationTuples.last._2
+                    (seqActivationId, seqActivation, atomicActionCnt)
+                } andThen {
+                    case Success((_, id, _)) => seqActivationPromise.success(id)
+                    case Failure(t)          => seqActivationPromise.failure(t)
+                }
+            }
+
+        // store result of sequence execution
+        // if seqActivation is defined, use it; otherwise create it (e.g., for non-blocking activations)
+        // the execution can reach here without a seqActivation due to non-blocking activations OR blocking activations that reach the blocking invoke timeout
+        // futureSeqResult should always be successful, if failed, there is an error
+        futureSeqResult flatMap { tuples => seqActivationPromise.future map { (tuples, _) } } onComplete {
+            case Success((wskActivationTuples, seqActivation)) =>
+                // all activations were successful
+                val activation = seqActivation getOrElse {
+                    val wskActivationEithers = wskActivationTuples.map(_._1)
+                    val end = Instant.now(Clock.systemUTC())
+                    // the response of the sequence is the response of the very last activation
+                    makeSequenceActivation(user, action, seqActivationId, wskActivationEithers, topmost, cause, start, end)
+                }
+                storeSequenceActivation(activation)
+            case Failure(t: Throwable) =>
+                // consider this whisk error
+                // TODO shall we attempt storing the activation if it exists or even inspect the futures?
+                // this should be a pretty serious whisk errror if it gets here
+                error(this, s"Sequence activation 'seqActivationId' failed: ${t.getMessage}")
+        }
+
+        response
+    }
+
+    /**
+     * Stores sequence activation to database.
+     */
+    private def storeSequenceActivation(activation: WhiskActivation)(implicit transid: TransactionId): Unit = {
+        WhiskActivation.put(activationStore, activation) onComplete {
+            case Success(id) => info(this, s"recorded activation")
+            case Failure(t)  => error(this, s"failed to record activation")
+        }
+    }
+
+    /**
+     * Creates an activation for a sequence.
+     */
+    private def makeSequenceActivation(
+        user: Identity,
+        action: WhiskAction,
+        activationId: ActivationId,
+        wskActivationEithers: Vector[Either[ActivationResponse, WhiskActivation]],
+        topmost: Boolean,
+        cause: Option[ActivationId],
+        start: Instant,
+        end: Instant): WhiskActivation = {
+
+        // extract all successful activations from the vector of activation eithers
+        val (right, left) = wskActivationEithers.span(_.isRight)
```

Comment:
Are you sure `span` is the right choice vs. `partition`?

From the scala doc:

> Splits this list into a prefix/suffix pair according to a predicate.

As I read that, `right` in your case is not guaranteed to only contain rights, which will result in an exception in the next line.